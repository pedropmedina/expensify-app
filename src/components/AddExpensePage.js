import React from 'react';

const AddExpensePage = () => {
	return <div>Add Expense Page</div>;
};

export default AddExpensePage;
