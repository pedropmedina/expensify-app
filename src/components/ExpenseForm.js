import React from 'react';

class ExpenseForm extends React.Component {
	state = { fields: { description: '', amount: '', note: '' } };

	onFieldChange = event => {
		const name = event.target.name;
		const val = event.target.value;
		const fields = this.state.fields;

		if (name === 'amount') {
			if (val.match(/^\d*(\.\d{0,2})?$/)) {
				fields[name] = val;
			}
		} else fields[name] = val;

		this.setState({ fields });
	};

	render() {
		const { description, amount, note } = this.state.fields;
		return (
			<form>
				<input
					type="text"
					name="description"
					value={description}
					placeholder="Description"
					autoFocus
					onChange={this.onFieldChange}
				/>
				<input
					type="text"
					name="amount"
					value={amount}
					placeholder="Amount"
					onChange={this.onFieldChange}
				/>
				<textarea
					name="note"
					value={note}
					placeholder="Add a note for  the expense"
					onChange={this.onFieldChange}
				/>
				<button>add expense</button>
			</form>
		);
	}
}

export default ExpenseForm;
