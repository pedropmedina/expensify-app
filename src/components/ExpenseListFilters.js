import React from 'react';
import { connect } from 'react-redux';

import { setTextFilter, sortByDate, sortByAmount } from '../actions/filters';

class ExpenseListFilters extends React.Component {
	state = { text: '' };

	onTextChange = event => {
		const text = event.target.value;
		this.setState({ text });
		this.props.setTextFilter(text);
	};

	onSortByChange = event => {
		const sortBy = event.target.value;
		if (sortBy === 'date') this.props.sortByDate();
		if (sortBy === 'amount') this.props.sortByAmount();
	};

	render() {
		return (
			<div>
				<input
					type="text"
					placeholder="filter by text"
					value={this.state.text}
					onChange={this.onTextChange}
				/>
				<select onChange={this.onSortByChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
			</div>
		);
	}
}

export default connect(
	undefined,
	{ setTextFilter, sortByAmount, sortByDate },
)(ExpenseListFilters);
