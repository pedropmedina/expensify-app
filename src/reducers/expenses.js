export const ADD_EXPENSE = 'ADD_EXPENSE';
export const EDIT_EXPENSE = 'EDIT_EXPENSE';
export const REMOVE_EXPENSE = 'REMOVE_EXPENSE';

const expensesReducerDefaultState = [];
const expenseReducer = (state = expensesReducerDefaultState, action) => {
	switch (action) {
		case ADD_EXPENSE:
			return [...state, action.expense];
		case EDIT_EXPENSE:
			return state.map(expense => {
				if (expense.id === action.id) {
					return { ...expense, ...action.updates };
				} else {
					return expense;
				}
			});
		case REMOVE_EXPENSE:
			return state.filter(({ id }) => id !== action.id);
		default:
			return state;
	}
};

export default expenseReducer;
