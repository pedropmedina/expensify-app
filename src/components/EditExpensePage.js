import React from 'react';
import { connect } from 'react-redux';

import { editExpense, startRemoveExpense } from '../actions/expenses';

import ExpenseForm from './ExpenseForm';

const EditExpensePage = props => {
	console.log(props);
	return (
		<div>
			<ExpenseForm
				{...props.expense}
				onSubmit={expense => {
					props.editExpense(props.match.params.id, expense);
					props.history.push('/');
				}}
			/>
			<button
				onClick={() => {
					props.startRemoveExpense({ id: props.match.params.id });
					props.history.push('/');
				}}
			>
				Remove
			</button>
		</div>
	);
};

const mapStateToProps = ({ expenses }, props) => ({
	expense: expenses.find(({ id }) => id === props.match.params.id),
});

export default connect(
	mapStateToProps,
	{ editExpense, startRemoveExpense },
)(EditExpensePage);
