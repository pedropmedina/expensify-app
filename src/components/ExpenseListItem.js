import React from 'react';
import { connect } from 'react-redux';

import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, removeExpense }) => (
	<div>
		<span>
			{description} >>>>> {amount}
		</span>
		<button onClick={() => removeExpense({ id })}>Remove</button>
	</div>
);

export default connect(
	undefined,
	{ removeExpense },
)(ExpenseListItem);
