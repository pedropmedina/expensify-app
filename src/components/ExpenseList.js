import React from 'react';
import { connect } from 'react-redux';
import uuidv4 from 'uuid/v4';

import selectExpenses from '../selectors/expenses';

import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = props => {
	return (
		<div>
			{props.expenses.length === 0 ? (
				<p>No Expenses!</p>
			) : (
				props.expenses.map(expense => {
					return <ExpenseListItem key={uuidv4()} {...expense} />;
				})
			)}
		</div>
	);
};

const mapStateToProps = ({ expenses, filters }) => ({
	expenses: selectExpenses(expenses, filters),
	filters,
});

export default connect(mapStateToProps)(ExpenseList);
