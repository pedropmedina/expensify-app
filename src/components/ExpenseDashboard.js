import React from 'react';
import { connect } from 'react-redux';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

class ExpenseDashboard extends React.Component {
	render() {
		return (
			<React.Fragment>
				<ExpenseListFilters />
				<ExpenseList />
			</React.Fragment>
		);
	}
}

export default ExpenseDashboard;
