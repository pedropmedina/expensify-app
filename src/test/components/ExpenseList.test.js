import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseList } from '../../components/ExpenseList';

import expenses from '../fixtures/expenses';

jest.mock('uuid/v4', () => jest.fn(() => 1));

test('should render ExpenseList with expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={expenses} />);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
	const wrapper = shallow(<ExpenseList expenses={[]} />);
	expect(wrapper).toMatchSnapshot();
});
