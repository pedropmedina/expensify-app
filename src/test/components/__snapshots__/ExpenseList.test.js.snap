// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No Expenses!
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={1.95}
    createdAt={0}
    description="Rent"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={212.87}
    createdAt={-345600000}
    description="Car"
    id="2"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={1}
    createdAt={345600000}
    description="Groceries"
    id="3"
    key="1"
    note=""
  />
</div>
`;
